pkg/result/processors/sort_results.go

```diff
@@ -0,0 +1,177 @@
+package processors
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/golangci/golangci-lint/pkg/config"
+	"github.com/golangci/golangci-lint/pkg/result"
+)
+
+// Base propose of this functionality to sort results (issues)
+// produced by various linters by analizing code. We achieving this
+// by sorting results.Issues using processor step, and chain based
+// rules that can compare different properties of the Issues struct.
+
+var _ Processor = (*SortResults)(nil)
+
+type SortResults struct {
+	cmp Comparer
+	cfg *config.Config
+}
+
+func NewSortResults(cfg *config.Config) *SortResults {
+	// For sorting we are comparing (in next order): file names, line numbers,
+	// position, range numbers (if position not available), and finnally - giving up.
+	return &SortResults{
+		cmp: ByName{
+			next: ByLine{
+				next: ByColumn{},
+			},
+		},
+		cfg: cfg,
+	}
+}
+
+// Process is performing sorting of the result issues.
+func (sr SortResults) Process(issues []result.Issue) ([]result.Issue, error) {
+	if !sr.cfg.Output.SortResults {
+		return issues, nil
+	}
+
+	sort.Slice(issues, func(i, j int) bool {
+		return sr.cmp.Compare(&issues[i], &issues[j]) == Less
+	})
+
+	return issues, nil
+}
+
+func (sr SortResults) Name() string { return "sort_results" }
+func (sr SortResults) Finish()      {}
+
+type CompareResult int
+
+const (
+	Less CompareResult = iota - 1
+	Equal
+	Greater
+	None
+)
+
+func (c CompareResult) isNeutral() bool {
+	// return true if compare result is incomparable or equal.
+	return c == None || c == Equal
+}
+
+func (c CompareResult) String() string {
+	switch c {
+	case Less:
+		return "Less"
+	case Equal:
+		return "Equal"
+	case Greater:
+		return "Greater"
+	}
+
+	return "None"
+}
+
+// Comparer describe how to implement compare for two "issues" lexicographically
+type Comparer interface {
+	Compare(a, b *result.Issue) CompareResult
+	Next() Comparer
+}
+
+var (
+	_ Comparer = (*ByName)(nil)
+	_ Comparer = (*ByLine)(nil)
+	_ Comparer = (*ByColumn)(nil)
+	_ Comparer = (*ByRange)(nil)
+)
+
+type ByName struct{ next Comparer }
+
+func (cmp ByName) Next() Comparer { return cmp.next }
+func (cmp ByName) Compare(a, b *result.Issue) CompareResult {
+	var res CompareResult
+
+	if res = CompareResult(strings.Compare(a.FilePath(), b.FilePath())); !res.isNeutral() {
+		return res
+	} else if next := cmp.Next(); res.isNeutral() && next != nil {
```

Comment:
res.isNeutral()  is required ?

Reply:
I would keep it, its simplify if/else branches. Having `res.isNeutral()` vs `res == Less || res == Greater` and `!res.IsNeutral()` vs `res == Equal || res == None`, imho better for understanding, live width and general complexity. If you are insisting - I will change, but I  prefer not to.

Reply:
oh what I mean is that !res.isNeutral() will return right away in previous if. I am thinking about something like

```suggestion
	if res = CompareResult(strings.Compare(a.FilePath(), b.FilePath())); !res.isNeutral() {
		return res
	}
	
	if next := cmp.Next(); next != nil {
		return next.Compare(a, b)
	}
	return res
```

Reply:
damn! my miss.